appnav/src/main/kotlin/io/element/android/appnav/di/DefaultSyncOrchestrator.kt

```diff
@@ -0,0 +1,148 @@
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
+ * Please see LICENSE files in the repository root for full details.
+ */
+
+package io.element.android.appnav.di
+
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import io.element.android.features.networkmonitor.api.NetworkMonitor
+import io.element.android.features.networkmonitor.api.NetworkStatus
+import io.element.android.libraries.core.coroutine.CoroutineDispatchers
+import io.element.android.libraries.matrix.api.MatrixClient
+import io.element.android.libraries.matrix.api.sync.SyncState
+import io.element.android.services.appnavstate.api.AppForegroundStateService
+import io.element.android.services.appnavstate.api.SyncOrchestrator
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import timber.log.Timber
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+class DefaultSyncOrchestrator @AssistedInject constructor(
+    @Assisted matrixClient: MatrixClient,
+    private val baseCoroutineScope: CoroutineScope = matrixClient.sessionCoroutineScope,
+    private val appForegroundStateService: AppForegroundStateService,
+    private val networkMonitor: NetworkMonitor,
+    private val dispatchers: CoroutineDispatchers,
+) : SyncOrchestrator {
+    @AssistedFactory
+    interface Factory {
+        fun create(matrixClient: MatrixClient): DefaultSyncOrchestrator
+    }
+
+    private val syncService = matrixClient.syncService()
+
+    private val initialSyncMutex = Mutex()
+
+    private var coroutineScope: CoroutineScope? = null
+
+    private val tag = "SyncOrchestrator"
+
+    private val started = AtomicBoolean(false)
+
+    /**
+     * Starting observing the app state and network state to start/stop the sync service.
+     *
+     * Before observing the state, a first attempt at starting the sync service will happen if it's not already running.
+     */
+    @OptIn(FlowPreview::class)
+    override fun start() {
+        if (!started.compareAndSet(false, true)) {
+            Timber.tag(tag).d("already started, exiting early")
+            return
+        }
+
+        Timber.tag(tag).d("start observing the app and network state")
+
+        if (syncService.syncState.value != SyncState.Running) {
+            Timber.tag(tag).d("initial startSync")
+            baseCoroutineScope.launch(dispatchers.io) {
+                try {
+                    initialSyncMutex.lock()
+                    syncService.startSync()
```

Comment:
Why do we need a first `startSync` here? The combine below is not enough to decide what to do depending on the status of all the observed flows?

Reply:
I was mimicking what we had before in `LoggedInFlowNode` for the airgapped case. After the changes in the SDK it 's probably not needed anymore.